chainstate/src/detail/chainstateref.rs

```diff
@@ -422,46 +427,393 @@ impl<'a, S: BlockchainStorageRead, O: OrphanBlocks> ChainstateRef<'a, S, O> {
-                            tx_id,
-                            block.get_id(),
-                        ))
-                    }
-                    None => txs_ids.insert(tx_id),
-                };
+    fn check_duplicate_txs(&self, block: &Block) -> Result<(), CheckBlockTransactionsError> {
+        // check duplicate transactions
+        let mut txs_ids = BTreeSet::new();
+        for tx in block.transactions() {
+            let tx_id = tx.get_id();
+            let already_in_tx_id = txs_ids.get(&tx_id);
+            match already_in_tx_id {
+                Some(_) => {
+                    return Err(CheckBlockTransactionsError::DuplicatedTransactionInBlock(
+                        tx_id,
+                        block.get_id(),
+                    ))
+                }
+                None => txs_ids.insert(tx_id),
+            };
+        }
+        Ok(())
+    }
+
+    fn check_issue_data(
+        &self,
+        token_ticker: &Vec<u8>,
+        amount_to_issue: &Amount,
+        number_of_decimals: &u8,
+        metadata_uri: &Vec<u8>,
+        tx_id: Id<Transaction>,
+        block_id: Id<Block>,
+    ) -> Result<(), CheckBlockTransactionsError> {
+        let mut is_correct_data = true;
+
+        // Check token name
+        is_correct_data = is_correct_data
+            && (token_ticker.len() <= TOKEN_MAX_TICKER_LEN)
+            && (!token_ticker.is_empty())
+            && String::from_utf8_lossy(token_ticker).is_ascii();
+        //TODO: Shall we have check for unique token name?
+
+        // Check amount
+        is_correct_data = is_correct_data && amount_to_issue != &Amount::from_atoms(0);
+
+        // Check decimals
+        is_correct_data = is_correct_data && number_of_decimals < &TOKEN_MAX_DEC_COUNT;
+
+        // Check URI
+        is_correct_data = is_correct_data
+            && metadata_uri.len() <= TOKEN_MAX_URI_LEN
+            && String::from_utf8_lossy(metadata_uri).is_ascii();
+
+        match is_correct_data {
+            true => Ok(()),
+            false => Err(CheckBlockTransactionsError::TokenIssueFail(tx_id, block_id)),
+        }
+    }
+
+    fn fetch(&self, outpoint: &OutPoint) -> Result<Transaction, CheckBlockTransactionsError> {
+        let tx_index = self
+            .db_tx
+            .get_mainchain_tx_index(&outpoint.tx_id())
+            .ok()
+            .flatten()
+            .ok_or(CheckBlockTransactionsError::FetchFail)?;
+        match tx_index.position() {
+            common::chain::SpendablePosition::Transaction(tx_pos) => {
+                let tx = self
+                    .db_tx
+                    .get_mainchain_tx_by_position(tx_pos)
+                    .ok()
+                    .flatten()
+                    .ok_or(CheckBlockTransactionsError::FetchFail)?;
+
+                Ok(tx)
+            }
+            common::chain::SpendablePosition::BlockReward(_) => {
+                todo!("Block reward in tokens is not allowed")
```

Comment:
Please don't use `todo!` unless it's something that will be done by the end of the PR, but return a proper error.

Reply:
I see, thanks!